source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/input/customcontrols/ValidatedCheckBox.java

```diff
@@ -0,0 +1,72 @@
+package io.adaptivecards.renderer.input.customcontrols;
+
+import android.content.Context;
+import android.graphics.Color;
+import android.graphics.PorterDuff;
+import android.support.v4.content.ContextCompat;
+import android.widget.CheckBox;
+import android.widget.LinearLayout;
+
+import io.adaptivecards.R;
+
+public class ValidatedCheckBox extends CheckBox implements IValidatedInputView
+{
+
+    public ValidatedCheckBox(Context context)
+    {
+        super(context);
+        setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
+        setBackground(ContextCompat.getDrawable(context, R.drawable.adaptive_choiceset_expanded_background));
+
+        m_isInvalid = false;
+    }
+
+    public ValidatedCheckBox(Context context, int errorColor)
+    {
+        this(context);
+        m_errorColor = errorColor;
+    }
+
+    /**
+     * Updates the rendered edit text to display the error visual cues
+     * @param isValid Value signalizing if input has valid value
+     */
+    @Override
+    public void setValidationResult(boolean isValid)
+    {
+        setInputInvalid(!isValid);
+
+        if (!isValid && getBackground() != null)
+        {
+            getBackground().setColorFilter(m_errorColor, PorterDuff.Mode.SRC_ATOP);
+        }
+        invalidate();
+    }
+
+    /**
+     * Refreshes the visual selector state for valid or invalid style
+     * @param isInvalid Value signalizing if the input has failed validation
+     */
+    public void setInputInvalid(boolean isInvalid)
+    {
+        m_isInvalid = isInvalid;
+        refreshDrawableState();
+    }
+
+    @Override
+    protected int[] onCreateDrawableState(int extraSpace)
+    {
+        final int[] drawableState = super.onCreateDrawableState(extraSpace + 1);
+        if (m_isInvalid)
+        {
+            mergeDrawableStates(drawableState, STATE_INPUT_INVALID);
+        }
+
+        return drawableState;
+    }
+
+    private boolean m_isInvalid = false;
+    private int m_errorColor = Color.TRANSPARENT;
+
+    private static final int[] STATE_INPUT_INVALID = {R.attr.adaptive_state_input_invalid};
```

Comment:
>R [](start = 54, length = 1)

really not liking the "R" in here -- what does it denote?

Reply:
as mentioned in a comment above, that's the android way of naming one of their APIs

https://developer.android.com/reference/android/R